### PR TITLE
Add a stateless test for the trailing empty chunk in `IMergingTransformBase`

### PR DESCRIPTION
**Pull request by groeneai**
### Changelog category (leave one):
- CI Fix or Improvement (changelog entry is not required)

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):

Add SQL-level regression coverage for the trailing empty chunk in `IMergingTransformBase::prepare`.

### Description

The reproducer in this test is @ tiandiwonder's, supplied in
https://github.com/ClickHouse/ClickHouse/pull/116805#issuecomment-5471983026. I am adding it to the
suite, with the comments trimmed and one of his two queries settled by measurement.

Related: https://github.com/ClickHouse/ClickHouse/pull/117048
Related: https://github.com/ClickHouse/ClickHouse/pull/116805

#117048 fixed `IMergingTransformBase::prepare`: an input that finishes its port in the same cycle it
delivers a trailing empty chunk had that chunk consumed by the merge. This test reaches that state
through a real read-in-order pipeline, where `BufferChunksTransform` is the producer and the tail of
each partition holds only empty arrays. #117048's description says the trigger "cannot be constructed
deterministically from SQL"; this shows that it can.

Verification: reverting only #117048's hunk on a debug build reddens the test **10 times out of 10**,
each time as `Logical error: 'max_rows > 0'` in `MergedData::rowsToInsertBeforeFlush` reached from
`MergingSortedAlgorithm::mergeBatchImpl` with the batch sorting queue, which aborts the server. On
master it passes 50/50 under randomized settings and 24/24 run against itself in parallel. The
pipeline-shape checks, one over each of the two runtime statements, also hold unchanged under
`allow_experimental_analyzer = 0` and `serialize_query_plan = 1`, the settings the old-analyzer and
distributed-plan test lanes install, and drop to 0 when buffering is disabled.

This does not make `gtest_merging_sorted_empty_chunk.cpp` redundant. `SortingStep::mergingSorted`
builds the merge only when there is more than one stream, and hard-codes
`SortingQueueStrategy::Batch`, so the gtest's single-input case and its default-strategy case are
both out of reach of this query; the gtest also builds its inputs by hand, where this test drives the
real pipeline.

The assertion is the only oracle this test carries: the empty cursor is heapified and compared before
that assertion is reached, so the release-visible row loss stays covered by the gtest's
`TwoInputsTrailingEmptyChunk`. `count` and `sum` in the second query detect lost or duplicated rows,
not reordering.

---
Workflow [[PR](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=117259&sha=latest&name_0=PR)]
Sync PR [[sync-upstream/pr/117259](https://github.com/search?q=head%3Async-upstream%2Fpr%2F117259+org%3AClickHouse+type%3Apr&type=pullrequests)]

### Version info
- Merged into: `26.9.1.473` (included in `26.9` and later)

### CODE REVIEW

**Comment by groeneai** [comment]
<details>
<summary>Internal second-model review (2 rounds) - 6 findings adjudicated</summary>

Before opening this PR I ran two rounds of independent cold review of the resulting code plus a
second-model review gate, and adjudicated every finding. Recorded here for auditability.

**Round 1**

| # | Severity | Finding | Verdict |
|---|---|---|---|
| ❌ 1 | major | The non-redundancy argument claimed `SortingQueueStrategy::Default` is not reachable from SQL. It is: `Interpreters/SortedBlocksWriter.cpp` builds a `MergingSortedTransform` with that strategy for a partial-merge join spilling to disk, and `Storages/MergeTree/MergeTask.cpp` selects it for background merges. Only the narrow claim holds: neither the default strategy nor the single-input shape is reachable through the read-in-order path this query drives. | AGREE, corrected in the description and the commit message |
| ❌ 2 | major | Nothing in the test asserted that buffering and the sorted merge are actually in the plan. `count` and `sum` are invariant under pipeline shape, so a planner change that stopped inserting `BufferChunksTransform` for this shape would leave the test green and silently vacuous. | AGREE, pipeline-shape assertion added with a negative control |
| ⚠️ 3 | nit | The commit message pinned a source line number, which rots. | AGREE, replaced by the symbol name |
| 💡 4 | nit | `read_in_order_two_level_merge_threshold` is randomized over `[0,100]` and the fixture holds 8 parts, so a low draw could reshape the read-in-order pipeline. | AGREE, measured. The low band is in fact unreachable here: a FINAL read returns from `spreadMarkRangesAmongStreamsFinal` before the branch that computes `need_preliminary_merge`. Swept `t` in {0,1,2,4,7,8,9,100} with the delivered value read back per row; the shape holds at every draw |

**Round 2**

| # | Severity | Finding | Verdict |
|---|---|---|---|
| ❌ 5 | major | The new shape assertion pins no lane settings, and plan shape is also moved by CI lane flags, which are runner options rather than randomized settings, so the 50-run randomized sweep cannot reach them. If a lane does not insert buffering, the new statement reds that lane. The nearest precedent in the suite, `04357_distributed_final_do_not_merge_partitions.sql`, uses the same predicate on the same query family and defends itself with `no-old-analyzer` plus pinned parallel-replicas and distributed-plan settings. | DISAGREE, refuted by measurement. The lanes install exactly `allow_experimental_analyzer=0` and `serialize_query_plan=1`. Running the whole file under each of those, and under parallel replicas, gives output byte-identical to the reference (same md5 on all four arms); the same probe returns `0` with buffering disabled, so the channel is sensitive rather than blind. Parallel replicas additionally cannot reach a FINAL read: that branch asserts `!is_parallel_reading_from_replicas` |
| 💡 6 | nit | PR metadata: relationship lines used shorthand rather than full GitHub URLs, and function names were written as `f()` rather than `f`. | AGREE, both fixed in this description |

On finding 2 the gate asked for an exact-count assertion on the four `BufferChunks` processors. I
used a boolean instead, because the exact-count form is measurably fragile: sibling
`03168_read_in_order_buffering_1` ran 267,342 times green across 44 check names in 60 days with
exactly one failure, and that failure was its `BufferChunks` line count moving from 1 to 2 in a
distributed-plan and parallel-replicas configuration. A `> 0` predicate absorbs both that and the
processor multiplicity `EXPLAIN PIPELINE` collapses onto one line.

Noted and not acted on: the second boolean in that statement is redundant, since `BufferChunks`
has a single construction site and the merge is added immediately after it in the same branch, so
the two columns cannot fall independently. It is kept because it is self-documenting and was
measured non-decorative (one partition, hence one stream, drops both).

</details>

**Comment by groeneai** [comment]
<details>
<summary><b>Pre-PR validation gate</b> (click to expand)</summary>

| # | Question | Answer |
|---|----------|--------|
| a | **Deterministic repro?** | Yes, 10/10. Revert #117048's hunk (`git checkout 49f3609beda^ -- src/Processors/Merges/IMergingTransform.cpp`), rebuild debug, run the test: `[ FAIL ] Reason: server died 32` every time. Not a rate. |
| b | **Root cause explained?** | Yes. `InputPort::pullData` caches `is_finished` with no `HAS_DATA` term, so the pre-fix guard `!input_chunk.hasRows() && !virtual_row && !input.isFinished()` let a rowless chunk from an already-finished port reach `state.has_input = true`. A zero-row cursor then entered the batch sorting queue and `rowsToInsertBeforeFlush(..., max_rows = 0, ...)` tripped `chassert(max_rows > 0)`. |
| c | **Fix matches root cause?** | No source fix in this PR: the fix is #117048, already on master. The coverage is validated against that fix's exact inverse hunk rather than against instrumentation markers. |
| d | **Test intent preserved / new tests added?** | New test added; nothing weakened or removed. Both regression queries were measured to reach the defect site on the reverted build, so neither arm is decorative. A third statement pins the pipeline shape the other two depend on, because `count`/`sum` are invariant under it; its own negative control (`read_in_order_use_buffering = 0`) reddens it on the reference diff. |
| e | **Both directions demonstrated?** | Yes, on the final shipped file. Red: 10/10 with the fix reverted (Build ID `2fa56d54...`), each run asserting the server's `buildId()` matches the on-disk binary. Green (Build ID `28531718...`): 50/50 under randomized settings, 24/24 self-parallel, 10/10 on the restored build. |
| f | **Fix is general across code paths?** | Carriers enumerated. `prepareInitializeInputs` already discarded rowless first chunks pre-fix; the other `IMergingTransform` subclasses share the fixed base `prepare`; `SortingQueueStrategy::Default` is out of reach of this query's pipeline because `SortingStep::mergingSorted` hard-codes `Batch` (other callers do reach it, `Interpreters/SortedBlocksWriter.cpp` and `Storages/MergeTree/MergeTask.cpp`, so it stays covered by #117048's gtest rather than being unreachable); the vertical-merge half of the seam (`ColumnGathererStream::onSourceExhausted`) is deliberately left to #116805. |
| g | **Fix generalizes across inputs (params/datatypes/wrappers)?** | A type-wrapper matrix does not apply: the fixed code is type-agnostic pipeline plumbing and this is a test file. The load-bearing axis is the settings matrix, and all six settings are pinned per-query so the runner's randomization cannot move them. |
| h | **Backward compatible? (maintainer-approved exception only)** | Yes, trivially. Test-only: no setting default, no serialization format, no `SettingsChangesHistory.cpp` entry. |
| i | **Invariants and contracts preserved?** | No source contract is touched. The test's own header comment states a standing property of the merge (it must not consume a rowless chunk), not a reference to the fix. |

</details>

Session id: cron:clickhouse-impl-slot-4:20260831-035500

**Comment by groeneai** [comment]
cc @davenger, could you review this? It adds SQL-level coverage for #117048, a read-in-order `FINAL` pipeline where `BufferChunksTransform` finishes its output port in the same cycle it pushes a trailing empty chunk, which before that fix reached `MergedData::rowsToInsertBeforeFlush` with `max_rows = 0`. Test-only; the last statement pins the pipeline shape with `EXPLAIN PIPELINE` so the coverage cannot go silently vacuous.

File: tests/queries/0_stateless/03168_read_in_order_buffering_3.sql [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        read_in_order_use_buffering = 1, max_threads = 1, max_block_size = 1
+)
+SETTINGS query_plan_remove_redundant_sorting = 0;
+
+-- The two queries above only reach the merge while the plan still puts buffering in front of a
+-- sorted merge, and their results are correct either way, so pin the shape rather than assume it.
+SET do_not_merge_across_partitions_select_final = 1, optimize_read_in_order = 1,
+    read_in_order_use_buffering = 1, max_threads = 1, max_block_size = 1;
+
+SELECT countIf(explain LIKE '%BufferChunks%') > 0, countIf(explain LIKE '%MergingSortedTransform%') > 0
```
[the remark this answers is not in the document]
Comment (groeneai) [reply tests/queries/0_stateless/03168_read_in_order_buffering_3.sql:55]: Agreed, and confirmed by measurement. Fixed in 377caf1ba8c6ff7.

The assertion alone would have made the test flaky, so this took two changes:

1. The wrapped statement's six load-bearing settings moved from its inner subquery to the outer
statement. `clickhouse-test` injects randomized settings as client options with
`--allow_repeated_settings`, and a client option beats a subquery-level `SETTINGS` clause, so a
randomized `optimize_read_in_order = 0`, drawn from `randint(0, 1)`, was already dropping
`BufferChunksTransform` from that statement's pipeline while `count` and `sum` kept returning
`3000 / 4498500 / 4498500`. That statement was blind today, not only after some future planner
change.

2. The shape assertion over that exact statement, as you asked.

Order matters. With the settings left on the inner subquery, adding the assertion fails under
`--client-option optimize_read_in_order=0` with `0 1` against `1 1`, while the same variant passes
with no client option. `EXPLAIN PIPELINE` of the wrapped statement on a debug build:

| settings placement | default | `--optimize_read_in_order 0` |
|---|---|---|
| inner subquery | `1 1` | `0 1` |
| outer statement | `1 1` | `1 1` |

The new line reddens both ways of losing the property: `read_in_order_use_buffering = 0` gives
`0 1`, and letting `query_plan_remove_redundant_sorting` default to 1 gives `0 0`, the case you
describe. The counts stay at their reference in both, which is why they cannot carry this.

`04489_virtual_row_disabled_with_final` was the right precedent: `c11b082a534ec09` is the same fix
on that file. Validated with 50 runs under default randomization, 8 concurrent runners times 3
runs, and 3 runs each under `allow_experimental_analyzer = 0`, `serialize_query_plan = 1` and
parallel replicas, no failures.
